rust/operator-binary/src/operations/graceful_shutdown.rs

```diff
@@ -0,0 +1,110 @@
+use indoc::formatdoc;
+use snafu::{ResultExt, Snafu};
+use stackable_druid_crd::security::DruidTlsSecurity;
+use stackable_druid_crd::DruidRole;
+use stackable_operator::k8s_openapi::api::core::v1::{ExecAction, LifecycleHandler};
+use stackable_operator::{
+    builder::{ContainerBuilder, PodBuilder},
+    time::Duration,
+};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Failed to set terminationGracePeriod"))]
+    SetTerminationGracePeriod {
+        source: stackable_operator::builder::pod::Error,
+    },
+}
+
+pub fn add_graceful_shutdown_config(
+    role: &DruidRole,
+    tls_security: &DruidTlsSecurity,
+    graceful_shutdown_timeout: Option<Duration>,
+    pod_builder: &mut PodBuilder,
+    druid_builder: &mut ContainerBuilder,
+) -> Result<(), Error> {
+    // This must be always set by the merge mechanism, as we provide a default value,
+    // users can not disable graceful shutdown.
+    if let Some(termination_grace_period) = graceful_shutdown_timeout {
+        match role {
+            DruidRole::Coordinator
+            | DruidRole::Broker
+            | DruidRole::Historical
+            | DruidRole::Router => {
+                pod_builder
+                    .termination_grace_period(&termination_grace_period)
+                    .context(SetTerminationGracePeriodSnafu)?;
+            }
+            DruidRole::MiddleManager => {
+                pod_builder
+                    .termination_grace_period(&termination_grace_period)
+                    .context(SetTerminationGracePeriodSnafu)?;
+
+                let (protocol, port) = if tls_security.tls_enabled() {
+                    ("https", role.get_https_port())
+                } else {
+                    ("http", role.get_http_port())
+                };
+
+                let middle_manager_host = format!("{protocol}://127.0.0.1:{port}");
+                let debug_message =
+                    "$(date --utc +%FT%T,%3N) INFO [stackable_lifecycle_pre_stop] -";
+                let sleep_interval = 2;
+
+                // The middle manager can be terminated gracefully by disabling it, meaning
+                // the overlord will not send any new tasks and it will terminate after
+                // all tasks are finished or the termination grace period is exceeded.
+                // See: https://druid.apache.org/docs/latest/operations/rolling-updates/#rolling-restart-graceful-termination-based
+                druid_builder.lifecycle_pre_stop(LifecycleHandler {
```

Comment:
a) Nit: would it be possible to format it to make it more readable in the pod yaml and b) can we maybe add some log output to the docs to show this actually happening? (maybe it does and I've missed it: see early comment about annotating the log output)

Reply:
Tried and failed :)

Reply:
a) no problem. b) As discussed: we'll this as-is to be consistent.